src/main/scala/cromwell/engine/workflow/SingleWorkflowRunnerActor.scala

```diff
@@ -27,104 +51,93 @@ object SingleWorkflowRunnerActor {
+    goto (RequestingMetadata)
   }
 
-  def receive = {
-    case workflowId: WorkflowId => subscribeToWorkflow(workflowId)
-    case outputs: Map[FullyQualifiedName@unchecked, WdlValue@unchecked] => outputOutputs(outputs)
-    case metadata: WorkflowMetadataResponse => outputMetadata(metadata) // pkg "webservice", but this is what WMA sends
-    case currentState: CurrentState[_] => log.debug(s"$tag: ignoring current state message: $currentState")
-    case Transition(_, _, state: WorkflowState) if state.isTerminal => handleTermination(state)
-    case Transition(_, _, state: WorkflowState) => log.info(s"$tag: transitioning to $state")
-    case Status.Failure(t) => handleFailure(t) // .pipeTo() transforms util.Failure() to Status.Failure()
-    case m => log.warning(s"$tag: received unexpected message: $m")
+  private def issueReply: State = {
+    self ! IssueReply
+    goto (Done)
   }
 
-  /**
-   * Logs the workflow completion, then either a) if the workflow completed successfully: requests the outputs, and
-   * later the metadata, or b) if not did not complete successfully, immediately requests the metadata.
-   * @param state The workflow termination state.
-   */
-  private def handleTermination(state: WorkflowState): Unit = {
-    log.info(s"$tag: workflow finished with status '$state'.")
-
-    // If this is a successful termination, retrieve & print out the outputs
-    if (state == WorkflowSucceeded) {
-      workflowManager ! WorkflowOutputs(id)
-    } else {
-      sendMetadataMessage()
-    }
+  when (NotStarted) {
+    case Event(RunWorkflow, data) =>
+      log.info(s"$tag: launching workflow")
+      workflowManager ! SubmitWorkflow(source)
+      goto (RunningWorkflow) using data.copy(replyTo = Option(sender()))
   }
 
-  /**
-   * Logs the errors and shuts down.
-   * @param t The error.
-   */
-  private def handleFailure(t: Throwable): Unit = {
-    log.error(t, s"$tag: ${t.getMessage}")
-    terminate()
+  when (RunningWorkflow) {
+    case Event(id: WorkflowId, data) =>
+      log.info(s"$tag: workflow ID UUID($id)")
+      workflowManager ! SubscribeToWorkflow(id)
+      stay using data.copy(id = Option(id))
+    case Event(Transition(_, _, WorkflowSucceeded), data) =>
+      workflowManager ! WorkflowOutputs(data.id.get)
+      goto(RequestingOutputs) using data.copy(terminalState = Option(WorkflowSucceeded))
+    case Event(Transition(_, _, state: WorkflowState), data) if state.isTerminal =>
+      // A terminal state that is not `WorkflowSucceeded` is a failure.
+      val updatedData = data.copy(terminalState = Option(state)).addFailure(s"Workflow ${data.id.get} transitioned to state $state")
+      // If there's an output path specified then request metadata, otherwise issue a reply to the original sender.
+      val nextState = if (metadataOutputPath.isDefined) requestMetadata else issueReply
+      nextState using updatedData
+    case Event(Transition(_, _, state), _) =>
```

Comment:
I think I've seen these in the unhandled blocks elsewhere, so this log gets written regardless